File: 0913_week5.cpp
// # 5주차_모음사전
// # https://programmers.co.kr/learn/courses/30/lessons/84512

#include <string>
#include <vector>
#include <assert.h>

using namespace std;

int solution(string word) {
    int n_length = word.length();

    int n_unit_count = ( ( ( 5 + 1 )*5 + 1 )*5 + 1 )*5 + 1; // 첫 번째 글자 기준 781, 156, 31, 6, 1
    for( auto c_letter : word ) {
        switch (c_letter) {
        case 'A':
            n_length += n_unit_count*0; break;
        case 'E':
            n_length += n_unit_count*1; break;
        case 'I':
            n_length += n_unit_count*2; break;
        case 'O':
            n_length += n_unit_count*3; break;
        case 'U':
            n_length += n_unit_count*4; break;
        default:
            assert("Wrong Letter");     break;
        }

        n_unit_count = (n_unit_count - 1) / 5;              // 한 글자씩 넘어갈수록 그에 맞게 사이즈 조절
    }

    return n_length;
}









// A
// AA
// AAA
// AAAA
// AAAAA
// AAAAE
// AAAAI
// AAAAO
// AAAAU
// AAAE
// AAAEA
// AAAEE
// AAAEI
// AAAEO
// AAAEU
// AAAI
// AAAIA
// AAAIE
// AAAII
// AAAIO
// AAAIU
// AAAO
// AAAOA
// AAAOE
// AAAOI
// AAAOO
// AAAOU
// AAAU

File: 0913/0913_week5.py
# 5주차_모음사전
# https://programmers.co.kr/learn/courses/30/lessons/84512

def _solution(word):
    dct_idx = { 
        'A':0, 
        'E':1, 
        'I':2, 
        'O':3, 
        'U':4,
    }
    
    answer = len(word)
    n_unit_count = (((5 + 1)*5 + 1)*5 + 1)*5 + 1
    for c_letter in word:
        answer += n_unit_count * dct_idx[c_letter]
        n_unit_count = (n_unit_count - 1) // 5
        
    return answer

    

if __name__ == '__main__':
    print(((((5 + 1)*5 + 1)*5 + 1)*5 + 1)*5)
    print(_solution('UUUUU'))